Code/GraphMol/Subgraphs/Subgraphs.cpp

```diff
@@ -575,8 +579,11 @@ PATH_TYPE findAtomEnvironmentOfRadiusN(const ROMol &mol, unsigned int radius,
 
         // add the next set of neighbors:
         int oAtom = mol.getBondWithIdx(bondIdx)->getOtherAtomIdx(startAtom);
+        if (atomMap.find(oAtom) == atomMap.end()) { atomMap[oAtom] = i + 1; }
+        else { atomMap[oAtom] = std::min(atomMap[oAtom], i + 1); }
+
         boost::tie(beg, end) = mol.getAtomBonds(mol.getAtomWithIdx(oAtom));
-        while (beg != end) {
+        while ((beg != end) && (i != radius - 1)) {
```

Comment:
I doubt that this is going to have any noticeable affect on performance (if you've done measurements and it does make a difference, please share the data and correct me though!), but it does create something that every other developer looking at this code is going to have to read and think about.
Please remove this complication.

Reply:
I think it reduce the work of bond retrievement at the `radius` distance. The idea is that we try to prevent bond retrieveing in this distance because we would never use it

Reply:
This is efficient if, let's say the last distance has large branching factor with on large bond

Reply:
I will add the doc to clarify this case. I think this optimization will worked well though, do you think, I should cast the loop to be from  `for (i = 0; i < radius; ++i)` to `for (i = 1 i <= radius; ++i)`. (At the end, it would be `i ~= radius + 1` instead) ?

Reply:
I understand what the code is intended to do, and I'm sure it does have some impact on run time. But I don't believe that it will actually make a significant difference and it does make the code more difficult to read/think about.
Please wait and don't make any more changes until after you have seen the PR I am going to make against your branch with my full set of suggested changes though.

Reply:
> I understand what the code is intended to do, and I'm sure it does have some impact on run time. But I don't believe that it will actually make a significant difference and it does make the code more difficult to read/think about. Please wait and don't make any more changes until after you have seen the PR I am going to make against your branch with my full set of suggested changes though.

Yes, I have tried to reproduce the algorithm in Python below, the performance shown that we are getting significant performance, especially when `radius=2` or `radius=3`.

Reply:
> Yes, I have tried to reproduce the algorithm in Python below, the performance shown that we are getting significant performance, especially when `radius=2` or `radius=3`.

you can't compare execution time in python to execution time in C++. The characteristics of the two languages are just too different.

Please take a look at the suggestions that I made in my PR... they include a form of this optimization, but in an easier-to-read way